File: src/ts/session.ts
// The drag session state machine — a pure apply(snapshot, input) → snapshot
// specified by docs/DESIGN.md §Session and proven by the shared trace corpus.
import { validateEnvelope, type DndDropResult, type DndEnvelope, type DndOperation, type ValidationOptions } from "./codec.js";
import { evaluatePolicy, validatePolicy, type DndDropPolicy, type DndRejectCode } from "./policy.js";

export type DndSessionState = "idle" | "dragging" | "over-target" | "dropped" | "cancelled";
export type DndSessionInputKind = "start" | "enter" | "leave" | "drop" | "cancel" | "end";

export interface DndSessionInput {
  kind: DndSessionInputKind;
  /** Required for `start`. */
  envelope?: DndEnvelope;
  /** Required for `enter`, `leave` and `drop`. */
  targetId?: string;
  /** Required for `enter`. */
  policy?: DndDropPolicy;
  /** Optional for `enter`. */
  preferredOperation?: DndOperation;
}

export interface DndSessionSnapshot {
  state: DndSessionState;
  dragId?: string;
  targetId?: string;
  operation?: DndOperation;
  errorCode?: DndRejectCode;
}

export interface DndSessionTrace {
  id: string;
  description?: string;
  inputs: DndSessionInput[];
  expected: DndSessionSnapshot[];
}

export const IDLE_SNAPSHOT: Readonly<DndSessionSnapshot> = Object.freeze({ state: "idle" });

export function isTerminal(state: DndSessionState): boolean {
  return state === "dropped" || state === "cancelled";
}

/** Input constructors. */
export const inputs = {
  start: (envelope: DndEnvelope): DndSessionInput => ({ kind: "start", envelope }),
  enter: (policy: DndDropPolicy, preferredOperation?: DndOperation): DndSessionInput =>
    preferredOperation ? { kind: "enter", targetId: policy.targetId, policy, preferredOperation } : { kind: "enter", targetId: policy.targetId, policy },
  leave: (targetId: string): DndSessionInput => ({ kind: "leave", targetId }),
  drop: (targetId: string): DndSessionInput => ({ kind: "drop", targetId }),
  cancel: (): DndSessionInput => ({ kind: "cancel" }),
  end: (): DndSessionInput => ({ kind: "end" }),
} as const;

interface SnapshotFields {
  dragId?: string | undefined;
  targetId?: string | undefined;
  operation?: DndOperation | undefined;
  errorCode?: DndRejectCode | undefined;
}

function snapshot(state: DndSessionState, fields: SnapshotFields = {}): DndSessionSnapshot {
  const out: DndSessionSnapshot = { state };
  if (fields.dragId !== undefined) out.dragId = fields.dragId;
  if (fields.targetId !== undefined) out.targetId = fields.targetId;
  if (fields.operation !== undefined) out.operation = fields.operation;
  if (fields.errorCode !== undefined) out.errorCode = fields.errorCode;
  return out;
}

/** The terminal DndDropResult for a snapshot, or null while the session runs. */
export function resultOf(s: DndSessionSnapshot): DndDropResult | null {
  if (s.dragId === undefined) return null;
  if (s.state === "dropped") {
    const result: DndDropResult = { dragId: s.dragId, accepted: true };
    if (s.operation !== undefined) result.operation = s.operation;
    if (s.targetId !== undefined) result.targetId = s.targetId;
    return result;
  }
  if (s.state === "cancelled") {
    const result: DndDropResult = { dragId: s.dragId, accepted: false };
    if (s.targetId !== undefined) result.targetId = s.targetId;
    if (s.errorCode !== undefined) result.errorCode = s.errorCode;
    return result;
  }
  return null;
}

export type SessionListener = (snapshot: DndSessionSnapshot, input: DndSessionInput) => void;

/** One drag session. Hosts keep one per drag source (or one global one). */
export class DndSession {
  #snapshot: DndSessionSnapshot = { state: "idle" };
  #envelope: DndEnvelope | null = null;
  readonly #options: ValidationOptions;
  readonly #listeners = new Set<SessionListener>();

  constructor(options: ValidationOptions = {}) {
    this.#options = options;
  }

  get snapshot(): DndSessionSnapshot {
    return this.#snapshot;
  }

  /** The envelope of the running (or just finished) session. */
  get envelope(): DndEnvelope | null {
    return this.#envelope;
  }

  get result(): DndDropResult | null {
    return resultOf(this.#snapshot);
  }

  subscribe(listener: SessionListener): () => void {
    this.#listeners.add(listener);
    return () => this.#listeners.delete(listener);
  }

  /** Apply one input and return the new snapshot. Malformed inputs are ignored. */
  apply(input: DndSessionInput): DndSessionSnapshot {
    this.#snapshot = this.#next(input);
    for (const listener of this.#listeners) listener(this.#snapshot, input);
    return this.#snapshot;
  }

  #next(input: DndSessionInput): DndSessionSnapshot {
    const current = this.#snapshot;
    if (input.kind === "start") {
      let envelope: DndEnvelope | null = null;
      try {
        envelope = input.envelope === undefined ? null : validateEnvelope(input.envelope, this.#options);
      } catch {
        envelope = null;
      }
      this.#envelope = envelope;
      return envelope ? snapshot("dragging", { dragId: envelope.dragId }) : snapshot("idle", { errorCode: "invalid-envelope" });
    }
    if (current.state === "idle" || isTerminal(current.state)) return current;
    switch (input.kind) {
      case "enter": {
        if (!input.policy || !this.#envelope) return current;
        let policy: DndDropPolicy;
        try {
          policy = validatePolicy(input.policy);
        } catch {
          return current;
        }
        const targetId = input.targetId ?? policy.targetId;
        const verdict = evaluatePolicy(this.#envelope, policy, input.preferredOperation);
        return verdict.accepted
          ? snapshot("over-target", { dragId: current.dragId, targetId, operation: verdict.operation })
          : snapshot("dragging", { dragId: current.dragId, targetId, errorCode: verdict.errorCode });
      }
      case "leave":
        return input.targetId !== undefined && current.targetId === input.targetId ? snapshot("dragging", { dragId: current.dragId }) : current;
      case "drop": {
        const cancelled = (errorCode: DndRejectCode): DndSessionSnapshot =>
          snapshot("cancelled", { dragId: current.dragId, targetId: input.targetId, errorCode });
        if (current.targetId === undefined) return cancelled("no-active-target");
        if (input.targetId !== current.targetId) return cancelled("target-mismatch");
        if (current.state === "over-target") return snapshot("dropped", current);
        return cancelled(current.errorCode ?? "no-active-target");
      }
      case "cancel":
      case "end":
        return snapshot("cancelled", { dragId: current.dragId, errorCode: "cancelled" });
      default:
        return current;
    }
  }
}

export interface TraceDivergence {
  traceId: string;
  step: number;
  expected?: DndSessionSnapshot;
  actual?: DndSessionSnapshot;
}

function sameSnapshot(a: DndSessionSnapshot, b: DndSessionSnapshot): boolean {
  return a.state === b.state && a.dragId === b.dragId && a.targetId === b.targetId && a.operation === b.operation && a.errorCode === b.errorCode;
}

/** Replay a trace from the idle state; returns the first divergence or null. */
export function replayTrace(trace: DndSessionTrace): TraceDivergence | null {
  if (trace.inputs.length !== trace.expected.length) {
    return { traceId: trace.id, step: Math.min(trace.inputs.length, trace.expected.length) };
  }
  const session = new DndSession();
  for (let step = 0; step < trace.inputs.length; step += 1) {
    const actual = session.apply(trace.inputs[step]!);
    const expected = trace.expected[step]!;
    if (!sameSnapshot(actual, expected)) return { traceId: trace.id, step, expected, actual };
  }
  return null;
}

File: src/rust/src/session.rs
//! The drag session state machine — a pure `apply(snapshot, input) → snapshot`
//! specified by docs/DESIGN.md §Session and proven by the shared trace corpus
//! under `contracts/instances/DndSessionTrace/valid/`.

use crate::envelope::{DndDropResult, DndEnvelope, DndError, DndOperation, ValidationOptions};
use crate::policy::{evaluate_policy, DndDropPolicy, DndRejectCode};
use crate::wire;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DndSessionState {
    Idle,
    Dragging,
    OverTarget,
    Dropped,
    Cancelled,
}

impl DndSessionState {
    pub const fn is_terminal(self) -> bool {
        matches!(self, DndSessionState::Dropped | DndSessionState::Cancelled)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum DndSessionInputKind {
    Start,
    Enter,
    Leave,
    Drop,
    Cancel,
    End,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DndSessionInput {
    pub kind: DndSessionInputKind,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub envelope: Option<DndEnvelope>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub target_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub policy: Option<DndDropPolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub preferred_operation: Option<DndOperation>,
}

impl DndSessionInput {
    fn bare(kind: DndSessionInputKind) -> Self {
        Self {
            kind,
            envelope: None,
            target_id: None,
            policy: None,
            preferred_operation: None,
        }
    }

    pub fn start(envelope: DndEnvelope) -> Self {
        Self {
            envelope: Some(envelope),
            ..Self::bare(DndSessionInputKind::Start)
        }
    }

    pub fn enter(policy: DndDropPolicy, preferred: Option<DndOperation>) -> Self {
        Self {
            target_id: Some(policy.target_id.clone()),
            policy: Some(policy),
            preferred_operation: preferred,
            ..Self::bare(DndSessionInputKind::Enter)
        }
    }

    pub fn leave(target_id: impl Into<String>) -> Self {
        Self {
            target_id: Some(target_id.into()),
            ..Self::bare(DndSessionInputKind::Leave)
        }
    }

    pub fn drop(target_id: impl Into<String>) -> Self {
        Self {
            target_id: Some(target_id.into()),
            ..Self::bare(DndSessionInputKind::Drop)
        }
    }

    pub fn cancel() -> Self {
        Self::bare(DndSessionInputKind::Cancel)
    }

    pub fn end() -> Self {
        Self::bare(DndSessionInputKind::End)
    }

    /// The structural rules both schema authorities check for an input. An
    /// embedded envelope is checked structurally only (its protocol version
    /// is the session's decision at `start`).
    pub fn structural(&self) -> Result<(), DndError> {
        if let Some(envelope) = &self.envelope {
            envelope.structural()?;
        }
        wire::check_opt_safe_id(self.target_id.as_deref(), "targetId")?;
        if let Some(policy) = &self.policy {
            policy.structural()?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DndSessionSnapshot {
    pub state: DndSessionState,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub drag_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub target_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub operation: Option<DndOperation>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error_code: Option<DndRejectCode>,
}

impl DndSessionSnapshot {
    pub const IDLE: DndSessionSnapshot = DndSessionSnapshot {
        state: DndSessionState::Idle,
        drag_id: None,
        target_id: None,
        operation: None,
        error_code: None,
    };

    fn dragging(drag_id: &str) -> Self {
        Self {
            state: DndSessionState::Dragging,
            drag_id: Some(drag_id.to_owned()),
            ..Self::IDLE
        }
    }

    /// True while the pointer is over a target that accepts the payload.
    pub fn is_over_accepting_target(&self) -> bool {
        self.state == DndSessionState::OverTarget
    }

    /// The final outcome once the session is terminal.
    pub fn result(&self) -> Option<DndDropResult> {
        let drag_id = self.drag_id.clone()?;
        match self.state {
            DndSessionState::Dropped => Some(DndDropResult {
                drag_id,
                accepted: true,
                operation: self.operation,
                target_id: self.target_id.clone(),
                error_code: None,
            }),
            DndSessionState::Cancelled => Some(DndDropResult {
                drag_id,
                accepted: false,
                operation: None,
                target_id: self.target_id.clone(),
                error_code: self.error_code.map(|code| code.wire().to_owned()),
            }),
            _ => None,
        }
    }
}

impl DndSessionSnapshot {
    pub fn structural(&self) -> Result<(), DndError> {
        wire::check_opt_safe_id(self.drag_id.as_deref(), "dragId")?;
        wire::check_opt_safe_id(self.target_id.as_deref(), "targetId")
    }
}

impl Default for DndSessionSnapshot {
    fn default() -> Self {
        Self::IDLE
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DndSessionTrace {
    pub id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    pub inputs: Vec<DndSessionInput>,
    pub expected: Vec<DndSessionSnapshot>,
}

impl DndSessionTrace {
    pub fn structural(&self) -> Result<(), DndError> {
        if !wire::is_trace_id(&self.id) {
            return Err(DndError(
                "trace id must match ^[a-z0-9][a-z0-9._-]{0,127}$".into(),
            ));
        }
        if self
            .description
            .as_deref()
            .is_some_and(|d| d.chars().count() > wire::TRACE_DESCRIPTION_MAX)
        {
            return Err(DndError("description exceeds 512 characters".into()));
        }
        wire::check_len(self.inputs.len(), 1, wire::TRACE_STEPS_MAX, "inputs")?;
        wire::check_len(self.expected.len(), 1, wire::TRACE_STEPS_MAX, "expected")?;
        if self.inputs.len() != self.expected.len() {
            return Err(DndError(
                "trace inputs and expected must have the same length".into(),
            ));
        }
        for input in &self.inputs {
            input.structural()?;
        }
        for snapshot in &self.expected {
            snapshot.structural()?;
        }
        Ok(())
    }
}

/// One drag session. Hosts keep one per drag source (or one global one) and
/// feed it native events translated to [`DndSessionInput`]s.
#[derive(Debug, Clone, Default)]
pub struct DndSession {
    snapshot: DndSessionSnapshot,
    envelope: Option<DndEnvelope>,
    options: ValidationOptions,
}

impl DndSession {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_options(options: ValidationOptions) -> Self {
        Self {
            options,
            ..Self::default()
        }
    }

    pub fn snapshot(&self) -> &DndSessionSnapshot {
        &self.snapshot
    }

    /// The envelope of the running (or just finished) session.
    pub fn envelope(&self) -> Option<&DndEnvelope> {
        self.envelope.as_ref()
    }

    /// Apply one input and return the new snapshot. Malformed inputs (an
    /// `enter` without a policy, a `leave` without a target) are ignored.
    pub fn apply(&mut self, input: &DndSessionInput) -> &DndSessionSnapshot {
        let next = self.next(input);
        self.snapshot = next;
        &self.snapshot
    }

    fn next(&mut self, input: &DndSessionInput) -> DndSessionSnapshot {
        let current = &self.snapshot;
        match input.kind {
            DndSessionInputKind::Start => {
                let valid = input
                    .envelope
                    .as_ref()
                    .filter(|envelope| envelope.validate(self.options).is_ok())
                    .cloned();
                match valid {
                    Some(envelope) => {
                        let snapshot = DndSessionSnapshot::dragging(&envelope.drag_id);
                        self.envelope = Some(envelope);
                        snapshot
                    }
                    None => {
                        self.envelope = None;
                        DndSessionSnapshot {
                            error_code: Some(DndRejectCode::InvalidEnvelope),
                            ..DndSessionSnapshot::IDLE
                        }
                    }
                }
            }
            _ if current.state == DndSessionState::Idle
                || current.state.is_terminal()
                || input.structural().is_err() =>
            {
                current.clone()
            }
            DndSessionInputKind::Enter => {
                let (Some(policy), Some(envelope)) =
                    (input.policy.as_ref(), self.envelope.as_ref())
                else {
                    return current.clone();
                };
                if policy.validate().is_err() {
                    return current.clone();
                }
                let target_id = input
                    .target_id
                    .clone()
                    .unwrap_or_else(|| policy.target_id.clone());
                let drag_id = current.drag_id.clone();
                match evaluate_policy(envelope, policy, input.preferred_operation) {
                    Ok(operation) => DndSessionSnapshot {
                        state: DndSessionState::OverTarget,
                        drag_id,
                        target_id: Some(target_id),
                        operation: Some(operation),
                        error_code: None,
                    },
                    Err(code) => DndSessionSnapshot {
                        state: DndSessionState::Dragging,
                        drag_id,
                        target_id: Some(target_id),
                        operation: None,
                        error_code: Some(code),
                    },
                }
            }
            DndSessionInputKind::Leave => match (&input.target_id, &current.target_id) {
                (Some(left), Some(active)) if left == active => DndSessionSnapshot {
                    state: DndSessionState::Dragging,
                    drag_id: current.drag_id.clone(),
                    ..DndSessionSnapshot::IDLE
                },
                _ => current.clone(),
            },
            DndSessionInputKind::Drop => {
                let dropped_on = input.target_id.clone();
                let cancelled = |code: DndRejectCode| DndSessionSnapshot {
                    state: DndSessionState::Cancelled,
                    drag_id: current.drag_id.clone(),
                    target_id: dropped_on.clone(),
                    operation: None,
                    error_code: Some(code),
                };
                match (&current.target_id, &dropped_on) {
                    (Some(active), Some(target)) if active == target => {
                        if current.state == DndSessionState::OverTarget {
                            DndSessionSnapshot {
                                state: DndSessionState::Dropped,
                                ..current.clone()
                            }
                        } else {
                            cancelled(current.error_code.unwrap_or(DndRejectCode::NoActiveTarget))
                        }
                    }
                    (Some(_), _) => cancelled(DndRejectCode::TargetMismatch),
                    (None, _) => cancelled(DndRejectCode::NoActiveTarget),
                }
            }
            DndSessionInputKind::Cancel | DndSessionInputKind::End => DndSessionSnapshot {
                state: DndSessionState::Cancelled,
                drag_id: current.drag_id.clone(),
                error_code: Some(DndRejectCode::Cancelled),
                ..DndSessionSnapshot::IDLE
            },
        }
    }

    /// Replay a trace from the idle state; returns the first divergence.
    pub fn replay(trace: &DndSessionTrace) -> Result<(), Box<TraceDivergence>> {
        if trace.inputs.len() != trace.expected.len() {
            return Err(Box::new(TraceDivergence {
                trace_id: trace.id.clone(),
                step: trace.inputs.len().min(trace.expected.len()),
                expected: None,
                actual: None,
            }));
        }
        let mut session = DndSession::new();
        for (step, (input, expected)) in trace.inputs.iter().zip(&trace.expected).enumerate() {
            let actual = session.apply(input).clone();
            if &actual != expected {
                return Err(Box::new(TraceDivergence {
                    trace_id: trace.id.clone(),
                    step,
                    expected: Some(expected.clone()),
                    actual: Some(actual),
                }));
            }
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TraceDivergence {
    pub trace_id: String,
    pub step: usize,
    pub expected: Option<DndSessionSnapshot>,
    pub actual: Option<DndSessionSnapshot>,
}

impl std::fmt::Display for TraceDivergence {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "trace {} diverged at step {}: expected {:?}, got {:?}",
            self.trace_id, self.step, self.expected, self.actual
        )
    }
}

impl std::error::Error for TraceDivergence {}
